src/apps/applib/lib/utility.c

```diff
@@ -177,6 +177,122 @@ int readBoundary(FILE* f, GeoBoundary* b) {
     return 0;
 }
 
+/**
+ * Move nonzero elements to the front of array `a` of length `n`.
+ *
+ * Loop invariant: Everything *before* `i` or *after* `j` is "done".
+ * Move `i` and `j` inwards until they equal, and exit.
+ * You can move `i` forward until there's a zero in front of it.
+ * You can move `j` backward until there's a nonzero to the left of it.
+ * Anything to the right of `j` is "junk" that can be reallocated.
+ *
+ * Before:
+ *   | a | b | 0 | c | d | ... |
+ *           ^           ^
+ *           i           j
+ * After:
+ *   | a | b | d | c | d | ... |
+ *           ^       ^
+ *           i       j
+ *
+ * todo: should this function be in the public API?
+ * todo: add tests for this function
+ *
+ * @param   a  H3Index array to whose elements will be moved
+ * @param   n  length of the input array
+ * @return     number of nonzero elements (length of new array); can reallocate
+ *             memory after this point
+ */
+size_t packNonzeros(H3Index* a, size_t n) {
+    size_t i = 0;
+    size_t j = n;
+
+    while (i < j) {
+        // move j to the left until the first nonzero
+        if (a[j - 1] == 0) {
+            j -= 1;
+            continue;
+        }
+
+        // move i to the right until the first zero
+        if (a[i] != 0) {
+            i += 1;
+            continue;
+        }
+
+        // if we get to this point, we know:
+        // a[i] == 0
+        // a[j-1] != 0
+        // i < j
+        // so we can swap! (actually, move a[j-1] -> a[i])
+        a[i] = a[j - 1];
+        j -= 1;
+    }
+
+    return i;
+}
+
+/**
+ * Array of all cells at a given resolution.
+ *
+ * @param   res  resolution
+ *
+ * @return       array of H3 cells at resolution res
+ */
+H3Index* getCellsAtRes(int res) {
+    int num0 = H3_EXPORT(res0IndexCount)();
+    H3Index* cells0 = calloc(num0, sizeof(H3Index));
+    H3_EXPORT(getRes0Indexes)(cells0);
+
+    int numRes = H3_EXPORT(maxUncompactSize)(cells0, num0, res);
+
+    H3Index* cellsRes = calloc(numRes, sizeof(H3Index));
+    H3_EXPORT(uncompact)(cells0, num0, cellsRes, numRes, res);
+
+    free(cells0);
+
+    numRes = packNonzeros(cellsRes, numRes);
+    cellsRes = realloc(cellsRes, numRes * sizeof(H3Index));
+
+    return cellsRes;
+}
+
+/**
+ * Apply callback to every cell for a given resolution, and sum the results.
+ */
+double mapSumAllCells_double(int res, double (*callback)(H3Index)) {
```

Comment:
Is there a reason to not base this on `iterateBaseCellIndexesAtRes` (allocate only memory for all indexes in each base cell, rather than the entire world?)

Reply:
With `iterateAllIndexesAtRes` the callback function doesn't return a value, so I wasn't sure how else to do the sum. Is there a standard way in C to achieve this "map a function over elements and sum the results"?

Reply:
We could refactor to combine these.

Reply:
What about something like this? This would keep the memory allocation to a minimum...

```c
void walk_children(H3Index h, int depth, void (*callback)(H3Index)){
    H3Index children[7] = {0};
    H3_EXPORT(h3ToChildren)(h, H3_GET_RESOLUTION(h)+1, children);

    if (depth == 1){
        for (int i = 0; i < 7; i++){
            if (children[i] != H3_NULL)
                (*callback)(children[i]);
        }
    } else { // depth > 1
        for (int i = 0; i < 7; i++){
            if (children[i] != H3_NULL)
                walk_children(children[i], depth-1, callback);
        }
    }
}
```

Reply:
> C would need to implement this with macros (or have an unused return value sometimes), so it would need refactoring

Yeah, that's what I was thinking. Just having an unused return value.

Reply:
shall we leave it as is for now, but make a ticket to ensure we revisit this in a future PR to clean up and refactor things?

Reply:
This sounds good

Reply:
Added https://github.com/uber/h3/issues/394